xarray/core/combine.py

```diff
@@ -369,37 +371,227 @@ def _auto_concat(datasets, dim=None, data_vars='all', coords='different'):
 _CONCAT_DIM_DEFAULT = '__infer_concat_dim__'
 
 
+def _infer_concat_order_from_positions(datasets, concat_dims):
+
+    combined_ids = OrderedDict(_infer_tile_ids_from_nested_list(datasets, ()))
+
+    tile_id, ds = list(combined_ids.items())[0]
+    n_dims = len(tile_id)
+    if concat_dims == _CONCAT_DIM_DEFAULT or concat_dims is None:
+        concat_dims = [concat_dims]*n_dims
+    else:
+        if len(concat_dims) != n_dims:
+            raise ValueError("concat_dims has length " + str(len(concat_dims))
+                             + " but the datasets passed are nested in a " +
+                             str(n_dims) + "-dimensional structure")
+
+    return combined_ids, concat_dims
+
+
+def _infer_tile_ids_from_nested_list(entry, current_pos):
+    """
+    Given a list of lists (of lists...) of objects, returns a iterator
+    which returns a tuple containing the index of each object in the nested
+    list structure as the key, and the object. This can then be called by the
+    dict constructor to create a dictionary of the objects organised byt their
+    position in the original nested list.
+
+    Recursively traverses the given structure, while keeping track of the
+    current position. Should work for any type of object which isn't a list.
+
+    Parameters
+    ----------
+    entry : list[list[obj, obj, ...]]
+        List of lists of arbitrary depth, containing objects in the order
+        they are to be concatenated.
+
+    Returns
+    -------
+    combined_tile_ids : dict[tuple(int, ...), obj]
+    """
+
+    if isinstance(entry, list):
+        for i, item in enumerate(entry):
+            for result in _infer_tile_ids_from_nested_list(item,
+                                                           current_pos + (i,)):
+                yield result
+    else:
+        yield current_pos, entry
+
+
+def _check_shape_tile_ids(combined_tile_ids):
+    tile_ids = combined_tile_ids.keys()
+
+    # Check all tuples are the same length
+    # i.e. check that all lists are nested to the same depth
+    nesting_depths = [len(id) for id in tile_ids]
```

Comment:
Use another name other than `id` here to avoid masking the builtin `id` (maybe `tile_id`?)